examples/hosts/iframe-host/src/inner.ts

```diff
@@ -24,26 +33,80 @@ async function getFluidObjectAndRender(container: Container, div: HTMLDivElement
     view.render(div, { display: "block" });
 }
 
-export async function runInner(divId: string) {
-    const div = document.getElementById(divId) as HTMLDivElement;
+async function loadFluidObject(
+    divId: string,
+    container: Container,
+) {
+    const componentDiv = document.getElementById(divId) as HTMLDivElement;
+    await getFluidObjectAndRender(container, componentDiv).catch(() => { });
+    // Handle the code upgrade scenario (which fires contextChanged)
+    container.on("contextChanged", (value) => {
+        getFluidObjectAndRender(container, componentDiv).catch(() => { });
+    });
+}
+
+async function loadContainer(
+    documentId: string,
+    createNew: boolean,
+    divId: string,
+): Promise<string> {
+    const documentServiceFactory = await InnerDocumentServiceFactory.create(innerPort);
+    const urlResolver = await InnerUrlResolver.create(innerPort);
 
-    const documentServiceFactory = await InnerDocumentServiceFactory.create();
     const module = { fluidExport: TodoContainer };
     const codeLoader = { load: async () => module };
 
     const loader = new Loader({
-        urlResolver: documentServiceFactory.urlResolver,
+        urlResolver,
         documentServiceFactory,
         codeLoader,
     });
 
-    const url = documentServiceFactory.resolvedUrl.url;
-    const documentId = decodeURIComponent(url.replace("fluid://localhost:3000/tinylicious/", ""));
-    const container = await loader.resolve({ url: documentId });
+    let container: Container;
 
-    await getFluidObjectAndRender(container, div).catch(() => { });
-    // Handle the code upgrade scenario (which fires contextChanged)
-    container.on("contextChanged", (value) => {
-        getFluidObjectAndRender(container, div).catch(() => { });
-    });
+    // TODO: drive new/existing creation entirely from outer
+    if (createNew) {
+        // We're not actually using the code proposal (our code loader always loads the same module regardless of the
+        // proposal), but the Container will only give us a NullRuntime if there's no proposal.  So we'll use a fake
+        // proposal.
+        container = await loader.createDetachedContainer({ package: "no-dynamic-package", config: {} });
+        // Caller is responsible for attaching the created container
+    } else {
+        // Request must be appropriate and parseable by resolver.
+        container = await loader.resolve({ url: documentId });
+        // If we didn't create the container properly, then it won't function correctly.  So we'll throw if we got a
+        // new container here, where we expect this to be loading an existing container.
+        // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+        if (!container.existing) {
+            throw new Error("Attempted to load a non-existing container");
+        }
+    }
+
+    await loadFluidObject(divId, container);
+    containers.set(container.id, container);
```

Comment:
this is fine for now, but long term this should live in the iframe driver. 
Basically, the outer should call some create method, maybe passing an iframe, which returns a container proxy to the detached container in the iframe, which can then have attach called on it.
Ideally this is just a usage example, and doesn't really contain scenario implementation.

Reply:
As in the iframe-driver layer is responsible for keeping track like this of containers created inside the iframe?  `containerProxy` class already allows the outer to do that, minus the passing an iframe part (this whole example is currently assuming a single iframe and should be refactored to support multiple).  I definitely just left this here for now because I wasn't sure where it should go and iframe-driver didn't seem quite right.

---
In reply to: [540465462](https://github.com/microsoft/FluidFramework/pull/4404#discussion_r540465462) [](ancestors = 540465462)